File: 059/spiral-matrix-ii.go
package main

import "fmt"

func generateMatrix(n int) [][]int {
	if n == 0 {
		return [][]int{}
	}

	res := make([][]int, n)
	for i := range res {
		res[i] = make([]int, n)
	}

	top, bottom, left, right := 0, n-1, 0, n-1
	num := 1
	for top <= bottom && left <= right {
		for j := left; j <= right; j++ {
			res[top][j] = num
			num++
		}
		top++

		for j := top; j <= bottom; j++ {
			res[j][right] = num
			num++
		}
		right--

		for j := right; j >= left; j-- {
			res[bottom][j] = num
			num++
		}
		bottom--

		for j := bottom; j >= top; j-- {
			res[j][left] = num
			num++
		}
		left++
	}
	return res
}

func main() {
	n := 3
	fmt.Println(generateMatrix(n))
}

File: 059/spiral-matrix-ii.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
    let res = [...Array(n)].map(e => Array(n));
    let val = 1, p = n;
    for (let i = 0; i < (n >> 1); i++, p -= 2) {
        for (let col = i; col < i+p; ++col)
            res[i][col] = val++;
        for (let row = i+1; row < i+p; ++row)
            res[row][i+p-1] = val++;
        for (let col = i+p-2; col >= i; --col)
            res[i+p-1][col] = val++;
        for (let row = i+p-2; row > i; --row)
            res[row][i] = val++;
    }
    if (n % 2 != 0) res[n>>1][n>>1] = val;
    return res;
};

let n = 3;
console.log(generateMatrix(n));
